File: julDir/recursive/collatzRecursive.jl
using Printf

# Result struct to hold input/output pair
struct Result
    input::UInt64
    output::UInt64
end

# CompareResults checks if a's output is larger or equal and input smaller than b's
# The outputs are flipped to provide descending sort
function compareResults(a::Result, b::Result)::Bool
    return !((a.output > b.output) || ((a.output == b.output) && (a.input < b.input)))
end

# CompareInputs checks if a's input is less than b's
function compareInputs(a::Result, b::Result)::Bool
    return a.input < b.input
end

# GetMinResult returns the minimum output value in topResults
function getMinResult(topResults::Vector{Result})::UInt64
    currentMin = topResults[1].output
    for i in 2:length(topResults)
        currentMin = min(currentMin, topResults[i].output)
    end
    return currentMin
end

# SortTopTen sorts the topResults based on sortType
# sortType true: highest output and smallest input -> lowest output
# sortType false: highest input -> smallest input
function sortTopTen(topResults::Vector{Result}, sortType::Bool)
    for j in 1:length(topResults)-1
        for i in j+1:length(topResults)
            swapItems = sortType ? compareResults(topResults[j], topResults[i]) : compareInputs(topResults[j], topResults[i])
            
            if swapItems
                topResults[j], topResults[i] = topResults[i], topResults[j]
            end
        end
    end
end

# UpdateTopResults uses x, y to create new Result (input x, output y),
# then checks if new Result's output already exists or is larger than
# current topResults. If the new Result meets the requirements, or
# topResults has less than 10, the new Result is added, then topResults is sorted
function updateTopResults(topResults::Vector{Result}, x::UInt64, y::UInt64)::Vector{Result}
    newResult = Result(x, y)
    newResultFits = false
    outputExistsAlready = false
    
    for i in 1:length(topResults)
        if topResults[i].output < newResult.output
            newResultFits = true
        end
        if (topResults[i].output == newResult.output) && (topResults[i].input < newResult.input)
            outputExistsAlready = true
        end
    end

    # Add the new item into topResults
    if (newResultFits || length(topResults) < 10) && !outputExistsAlready
        push!(topResults, newResult)
        sortTopTen(topResults, true)
    end

    # Keep only the top 10 results
    if length(topResults) > 10
        topResults = topResults[1:10]
    end

    return topResults
end

# PrintResults prints topResults based on seq length and input size
function printResults(topResults::Vector{Result})
    # Print top 10 based on result length
    println("Sorted based on sequence length")
    for i in 1:length(topResults)
        @printf "%20i%20i\n" topResults[i].input topResults[i].output
    end

    # Print top 10 based on input length
    sortTopTen(topResults, false)
    println("Sorted based on integer length")
    for i in 1:length(topResults)
        @printf "%20i%20i\n" topResults[i].input topResults[i].output
    end
end

# CollatzNumofSequences calculates the number of sequences it takes
# before the input = 1. Then returns the num of sequences
function collatzNumofSequences(input::UInt64, currStep::UInt64)::UInt64
    baseCase::UInt64 = 1
    if input <= baseCase
        return currStep
    end

    if input % 2 == 1
        return collatzNumofSequences((input * 3 + 1), (currStep + 1))
    else
        evenResult::UInt64 = (input / 2)
        return collatzNumofSequences(evenResult, (currStep + 1))
    end
end

# CollatzNumofSequencesPrintTopTen repeats CollatzNumofSequences for each number in the given range
# It stores the top ten results, then prints them at the end
function collatzNumofSequencesPrintTopTen(lowerLimit::UInt64, upperLimit::UInt64)
    topTen = Result[]
    currMin::UInt64 = 0
    
    while (upperLimit >= lowerLimit)
        result = collatzNumofSequences(lowerLimit, UInt64(0))
        
        if result > currMin || length(topTen) < 10
            topTen = updateTopResults(topTen, lowerLimit, result)
            currMin = getMinResult(topTen)
        end

        lowerLimit = lowerLimit + 1
    end
    
    printResults(topTen)
end

# Main function
function main()
    lowerLimit, upperLimit, inputLimit = 0, 0, 2100000001 # 2.1 billion
        # Check args for limits
        if length(ARGS) > 0
            lowerLimit = parse(UInt64, ARGS[1])
        end
        if length(ARGS) > 1
            upperLimit = parse(UInt64, ARGS[2])
        end
    
        # Check if bounds set properly
        if lowerLimit > inputLimit || upperLimit > inputLimit
            println("Input is too large, please use a num 0 - 2.1 billion")
            return
        end
    
        if lowerLimit > 0 && upperLimit == 0
            result = collatzNumofSequences(lowerLimit, (lowerLimit - lowerLimit)) 
            @printf "%20i%20i\n" lowerLimit result
        elseif lowerLimit > 0 && upperLimit > 0 && upperLimit > lowerLimit
            collatzNumofSequencesPrintTopTen(lowerLimit, upperLimit)
        end

        return
    end
    
    # Run the main function if the script is executed
    main()

File: goDir/recursive/collatzRecursive.go
package main

import (
  "fmt"
  "os"
  "strconv"
)

// Result struct to hold input/output pair
type Result struct {
  Input  uint64
  Output uint64
}

// CompareResults checks if a's output is larger or equal and input smaller than b's
// The outputs are flipped to provide descending sort
func CompareResults(a Result, b Result) bool {
  if a.Output > b.Output {
    return !true
  } else if a.Output == b.Output && a.Input < b.Input {
    return !true
  }
  return !false
}

// CompareInputs checks if a's input is less than b's
func CompareInputs(a Result, b Result) bool {
  return a.Input < b.Input
}

func GetMinResult(topResults []Result) uint64 {
	var currentMin uint64
	currentMin = topResults[0].Output

	for i := 0; i < len(topResults); i++ {
		if (topResults[i].Output < currentMin) {
			currentMin = topResults[i].Output
		}
	}

	return currentMin
}

// SortTopTen sorts the topResults based on sortType
// sortType true: highest output and smallest input -> lowest output
// sortType false: highest input -> smallest input
func SortTopTen(topResults []Result, sortType bool) {
  for j := 0; j < len(topResults)-1; j++ {
    for i := j + 1; i < len(topResults); i++ {
      var swapItems bool

      // Determine if swap is needed based on sortType
      // sortType true: highest output and smallest input -> lowest output
      // sortType false: highest input -> smallest input
      if sortType {
        swapItems = CompareResults(topResults[j], topResults[i])
      } else {
        swapItems = CompareInputs(topResults[j], topResults[i])
      }

      // Swap items if needed
      if swapItems {
        topResults[j], topResults[i] = topResults[i], topResults[j]
      }
    }
  }
}

// UpdateTopResults uses x, y to create new Result (input x, output y),
// then checks if new Result's output already exists or is larger than
// current topResults. If the new Result meets the requirements, or
// topResults has less than 10, the new Result is added, then topResults is sorted
func UpdateTopResults(topResults []Result, x uint64, y uint64) []Result {
  newResult := Result{x, y}
  var newResultFits, outputExistsAlready bool
  

  for i := 0; i < len(topResults); i++ {
    if topResults[i].Output < newResult.Output {
      newResultFits = true
    }
    if ((topResults[i].Output == newResult.Output) && topResults[i].Input < newResult.Input) {
      outputExistsAlready = true
    }
  }

  // Add the new item into topResults
  if (newResultFits || len(topResults) < 10) && !outputExistsAlready {
    newTR := append(topResults, newResult)
	topResults = newTR
    SortTopTen(topResults, true)
  }

  // Keep only the top 10 results
  if len(topResults) > 10 {
    topResults = topResults[:10]
  }

  return topResults
}

// PrintResults prints topResults based on seq length and input size
func PrintResults(topResults []Result) {
  // Print top 10 based on result length
  fmt.Println("Sorted based on sequence length")
  for i := 0; i < len(topResults); i++ {
    fmt.Printf("%20d%20d\n", topResults[i].Input, topResults[i].Output)
  }

  // Print top 10 based on input length
  SortTopTen(topResults, false)
  fmt.Println("Sorted based on integer length")
  for i := 0; i < len(topResults); i++ {
    fmt.Printf("%20d%20d\n", topResults[i].Input, topResults[i].Output)
  }
}

// CollatzNumofSequences calculates the number of sequences it takes
// before the input = 1. Then returns the num of sequences
func CollatzNumofSequences(input uint64, currStep uint64) uint64 {
  if (input == 1) {
	return currStep
  }

  if (input % 2 == 1) {
	return CollatzNumofSequences((input*3 +1), (currStep + 1))
  } else {
	return CollatzNumofSequences((input / 2), (currStep + 1))
  }

}

// CollatzNumofSequencesPrintTopTen repeats CollatzNumofSequences for each number in the given range
// It stores the top ten results, then prints them at the end
func CollatzNumofSequencesPrintTopTen(lowerLimit uint64, upperLimit uint64) {
  var topTen []Result
  var currMin uint64 // to skip any values below topTen min
  numCycles := 0

  for i := lowerLimit; i <= upperLimit; i++ {
    result := CollatzNumofSequences(i, 0)

	if result > currMin ||  numCycles < 10 {
      topTen = UpdateTopResults(topTen, i, result)
	  currMin = GetMinResult(topTen)
    }

	numCycles++
  }

  PrintResults(topTen)
}

// Main function
func main() {
  var lowerLimit, upperLimit, inputLimit uint64 = 0, 0, 2100000001 // 2.1 billion + 1

  // AbsoluteLimit is the largest possible uint in Go
  //var absoluteLimit uint64 = math.MaxUint64 // 18446744073709551615
  // AbsoluteOdd is the largest odd number that can be called within the collatz func
  //var absoluteOdd uint64 = ((absoluteLimit - 1) / 3) - 1 // 1,431,655,763

  // Check args for limits
  if len(os.Args) > 1 {
    lowerLimit, _ = strconv.ParseUint(os.Args[1], 10, 64)
  }
  if len(os.Args) == 3 {
    upperLimit, _ = strconv.ParseUint(os.Args[2], 10, 64)
  }

  // Check if bounds set properly
  if lowerLimit > inputLimit || upperLimit > inputLimit {
    fmt.Println("Input is too large, please use a num 0 - 2.1 billion")
    return
  }

  if lowerLimit > 0 && upperLimit == 0 {
    result := CollatzNumofSequences(lowerLimit, 0)
    fmt.Printf("%20d%20d\n", lowerLimit, result)
  } else if lowerLimit > 0 && upperLimit > 0 && upperLimit > lowerLimit {
    CollatzNumofSequencesPrintTopTen(lowerLimit, upperLimit)
  }
}
